### ISSUE TITLE
Domain for deposit signatures

### ISSUE THREAD

**Issue by paulhauner** [prysmaticlabs/prysm#3849]
Hey Prysm 👋

I'm working to build a genesis state from your deposit contract but all the signatures are failing on our end. After some digging, I suspect there's a mismatch between Prysm and the spec.

In `process_deposit`, the spec declares:

```python
# Note: Deposits are valid across forks, thus the deposit domain is retrieved directly from `compute_domain`.
domain = compute_domain(DOMAIN_DEPOSIT)
```

The second parameter to `compute_domain` is omitted, so it defaults to `Version()` which I understand to be `[0, 0, 0, 0]`.

In Prysm it seems you're passing through the `Fork` from the `BeaconState`:

https://github.com/prysmaticlabs/prysm/blob/9bf55e53e70bfda50cc82ef460a1470d89b90e54/beacon-chain/core/blocks/block_operations.go#L966-L972

Which in your testnet is set to this:

https://github.com/prysmaticlabs/prysm/blob/acf11262de3439ffa4c976565d59d0739c18ce29/shared/params/config.go#L236

So, as I understand it, the spec calls for `[0, 0, 0, 0]` but Prysm is using `[0, 0, 0, 2]`. If this turns out to be the case and you're passing all the v0.8.3 tests, it seems that the EF tests have a coverage gap. This seems feasible if all the tests assume that the `Fork` is `[0, 0, 0, 0]`.

**Comment by terencechain** [comment]
Ahh yes. thank you for pointing that out. We bumped up the fork version on every test net restarts for critical bug fixes. This is to disconnect with external peers that may no longer be compatible to sync with post critical bug fixes.

Not sure what's the best course forward. Is rolling back to fork version 0 a possibility? @prestonvanloon @rauljordan @nisdas

**Comment by paulhauner** [comment]
You're welcome!

I think you can keep bumping the fork version, but deposits need to stay on the 0 fork/domain to stick with the spec. All other non-deposit messages will still fail verification.

Interestingly, keeping the deposits on the 0 fork/domain means that you can replay deposits across different testnets. This doesn't immediately seem like a vuln though, there seems to be no gain to spending `x ETH` to place a deposit for a key that you don't own. In fact, this might even be a feature! It means we could seamlessly move validators from an existing testnet over to a new one, even if the testnet fork changes.

**Comment by terencechain** [comment]
> Interestingly, keeping the deposits on the 0 fork/domain means that you can replay deposits across different testnets.

I'm just happy when we are discovering new "features" along with this experiment ; )

**Comment by nisdas** [comment]
@paulhauner Actually for our tests our fork version is still 
```go
[]byte{0, 0, 0, 0}
```
so that is why we are still compliant with the spectests. The only time the fork version diverges 
is during runtime where we use our `demoConfig` to set our custom parameters. However for the purposes of running our test suite it chooses the default beacon config, which always has the fork version as 
```go
[]byte{0, 0, 0, 0}
```

**Comment by paulhauner** [comment]
@nisdas, makes sense. It would also break in the case where the fork version was incremented by a legit hard-fork, though.

**Comment by nisdas** [comment]
Not exactly, if there was a hard-fork, we would increment the fork version in the default config(Mainnet Config) . 
https://github.com/prysmaticlabs/prysm/blob/master/shared/params/config.go#L173.

So we would then use this fork version across all our tests, etc.

**Comment by paulhauner** [comment]
I mean the client would diverge from consensus at runtime because it would declare all deposit signatures invalid.

The value you linked me wouldn't be what we increment during a fork, it would be `fork.current_version`.

I open an issue to add a test case for this here: https://github.com/ethereum/eth2.0-specs/issues/1445

**Comment by nisdas** [comment]
Ah k. Yes in that case it would declare some deposit signatures as invalid. Processing deposit logs from the eth1 chain would be problematic in this case

**Comment by prestonvanloon** [comment]
What is the resolution for this issue?

**Comment by prestonvanloon** [comment]
I believe this was resolved by #3876. Please reopen if that is not the case.
